File: typescript/src/workshop/m6/s13_extractcomposite/step1/Marathon.ts
import { CompositeProgramItem } from './CompositeProgramItem.js';

/** Krok 1: pole children, add i children() podciągnięte do CompositeProgramItem. */
export class Marathon extends CompositeProgramItem {
  constructor(private readonly name: string) {
    super();
  }

  override minutes(): number {
    const children = this.children();
    let total = 0;
    for (const child of children) {
      total += child.minutes();
    }
    return children.length === 0 ? 0 : total + 15 * (children.length - 1);
  }

  override describe(): string {
    return `Maraton ${this.name} (${this.minutes()} min) `
      + `[${this.children().map((child) => child.describe()).join(', ')}]`;
  }
}

File: csharp/src/Training.Workshop/M6/S13ExtractComposite/Step1/Marathon.cs
namespace Training.Workshop.M6.S13ExtractComposite.Step1;

/// <summary>Krok 1: pole _children, Add i Children podciągnięte do CompositeProgramItem.</summary>
public sealed class Marathon : CompositeProgramItem
{
    private readonly string _name;

    public Marathon(string name)
    {
        _name = name;
    }

    public override int Minutes
    {
        get
        {
            var children = Children;
            var total = 0;
            foreach (var child in children)
            {
                total += child.Minutes;
            }
            return children.Count == 0 ? 0 : total + 15 * (children.Count - 1);
        }
    }

    public override string Describe()
    {
        return "Maraton " + _name + " (" + Minutes + " min) "
            + "[" + string.Join(", ", Children.Select(child => child.Describe())) + "]";
    }
}
